#**Import required libraries**

In [1]:
import pandas as pd
import numpy as np


    Pandas:
         - It used for data manipulation and analysis.
         - It provides two main structures: Series for one-dimensional data and DataFrames for two-dimensional data.
         - With Pandas, we can easily:
             - Read and write data from various formats
             - Handle missing values
             - Filter data
             - Group it for aggregation
             - Merge multiple datasets

    NumPy
        - It is focused on numerical computing.
        - It provides support for multi-dimensional arrays (ndarray) and offers a wide range of mathematical functions for array operations.
        - It excels at:
             - Linear algebra
             - Broadcasting
             - Random number generation

#**W1-GA1.1**

In [2]:
df= pd.read_csv("/content/drive/MyDrive/MLP/W1/V5 (1).csv")
df

Date  Year    Locality  Estimated Value  Sale Price  \
0     2009-01-02  2009   Waterbury         111440.0    185000.0   
1     2009-01-02  2009  Bridgeport         124670.0    150000.0   
2     2009-01-02  2009   Waterbury          55720.0    140000.0   
3     2009-01-02  2009  Bridgeport        4775276.0    272900.0   
4     2009-01-02  2009  Bridgeport         112351.0    210000.0   
...          ...   ...         ...              ...         ...   
9995  2022-09-30  2022  Bridgeport         153660.0    322000.0   
9996  2022-09-30  2022  Bridgeport         109820.0    340000.0   
9997  2022-09-30  2022   Waterbury          83210.0    170000.0   
9998  2022-09-30  2022         NaN         216420.0    415000.0   
9999  2022-09-30  2022     Norwalk         154920.0    196000.0   

           Property     Residential  num_rooms  num_bathrooms  carpet_area  \
0     Single Family  Detached House          3              3        996.0   
1        Two Family          Duplex          4              3       1241.0   
2     Single Family  Detached House          3              2        910.0   
3     Single Family  Detached House          3              1        971.0   
4                 ?  Detached House          3              2       1092.0   
...             ...             ...        ...            ...          ...   
9995              ?  Detached House          3              3        917.0   
9996              ?          Duplex          4              4       1416.0   
9997  Single Family  Detached House          3              1        964.0   
9998  Single Family  Detached House          3              3       1014.0   
9999  Single Family  Detached House          3              1        932.0   

      property_tax_rate   Face  
0              1.025953  South  
1              1.025953  South  
2              1.025953  South  
3              1.025953   East  
4              1.025953   East  
...                 ...    ...  
9995           1.003979  North  
9996           1.003979  North  
9997           1.003979  South  
9998           1.003979  North  
9999           1.003979   East  

[10000 rows x 12 columns]

In [3]:
df.shape

(10000, 12)

       shape attribute is used to get the dimensions of a NumPy array or a Pandas DataFrame.
       It returns a tuple representing the size of each dimension and helps to understand data dimensionality.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Date               10000 non-null  object 
 1   Year               10000 non-null  int64  
 2   Locality           8747 non-null   object 
 3   Estimated Value    8757 non-null   float64
 4   Sale Price         10000 non-null  float64
 5   Property           10000 non-null  object 
 6   Residential        10000 non-null  object 
 7   num_rooms          10000 non-null  int64  
 8   num_bathrooms      10000 non-null  int64  
 9   carpet_area        8771 non-null   float64
 10  property_tax_rate  10000 non-null  float64
 11  Face               10000 non-null  object 
dtypes: float64(4), int64(3), object(5)
memory usage: 937.6+ KB


**INSIGHTS**
- Total Entries (Rows): 10,000
- Number of Columns: 12
- Data Types: Some columns contain integer data (int64), some contain floating-point numbers (float64), and others contain strings/objects (object).
- Missing Values: Columns like "Locality," "Estimated Value," "Property," and "carpet_area" have some missing values, indicated by a non-null count less than 10,000.
- Memory Usage: The DataFrame uses around 937.6 KB of memory.

    info() is a quick and efficient way to get an overview of a DataFrame.
    Helps to identify data types, missing values, and the overall structure of the dataset.

###**Q1**
**How many unknown ("?) values are present in the dataset? Remove/Delete unknown ("?) values present in the dataset to make it null value.
Remove/Delete - means it will show NAN in place of "?"**

**Note: If there is no value present in the dataset it is represented as NAN(read pandas documentation for all the other ways to represent null values) in data**

In [5]:
count = (df == '?').sum()
count

Date                    0
Year                    0
Locality                0
Estimated Value         0
Sale Price              0
Property             1823
Residential             0
num_rooms               0
num_bathrooms           0
carpet_area             0
property_tax_rate       0
Face                    0
dtype: int64

In [6]:
unknown_count = (df == '?').sum().sum()
unknown_count

1823

    sum() method in Pandas is used to calculate the sum of values in a DataFrame or Series.
    It can be applied to numeric data types, and can specify the axis along which to perform the sum.

##**Q2**
**What is the shape of the dataset ?**


In [7]:
df.shape

(10000, 12)

##**Q3**
**What is the value present at the 692th indexed row and 0th indexed column in the data ?**


In [8]:
df=pd.DataFrame(df)

In [9]:
df.iloc[692,0]

'2009-11-16'

    iloc[] indexer in Pandas is used to access a DataFrame by its integer-location based indexing.
    It allows to select rows and columns based on their positions (integer indices) rather than their labels.

##**Q4**

**What is the value present at the 546th indexed row and 7th indexed column in the data ?**

In [10]:
df.iloc[567,7]

3

##**Q5**
**What are the unique values present in the Locality feature of the dataset?**

In [11]:
df['Locality'].unique()

array(['Waterbury', 'Bridgeport', 'Greenwich', 'Norwalk', nan,
       'Fairfield', 'West Hartford', 'Stamford'], dtype=object)

    unique() method in Pandas is used to extract the unique values from a Series or DataFrame column.
    It returns an array of the unique values, which is particularly useful for identifying distinct entries in data.

##**Q6**
**Which of the following features have missing(NaN) values present in the dataset? (Note: compute after removing "?")**

In [12]:
df=df.replace('?', np.nan)

    `replace()` method in Python is used to replace occurrences of a specified substring within a string with another substring.

    It has the following syntax:
           string.replace(old, new, count)
              - old: The substring you want to replace.
              - new: The substring to replace it with.
              - count: (Optional) The number of occurrences to replace. If not specified, it replaces all occurrences.

In [13]:
null_value = df.isnull()

     isnull() method in Pandas is used to detect missing values (NaN) in a DataFrame or Series.
     It returns a Boolean DataFrame or Series, indicating whether each value is NaN (True) or not (False).

In [14]:
null_value.sum()

Date                    0
Year                    0
Locality             1253
Estimated Value      1243
Sale Price              0
Property             1823
Residential             0
num_rooms               0
num_bathrooms           0
carpet_area          1229
property_tax_rate       0
Face                    0
dtype: int64

In [15]:
null_value.sum().sum()

5548

##**Q7**
**Which of the following feature has most missing(NaN) values present in the dataset? (Note: compute after removing "?")**

In [16]:
print(null_value.sum().idxmax(),null_value.sum().max())

Property 1823


    max() - finds the actual maximum value in  data.

    idxmax() - find the index (or row label) of where that maximum value occurs.

##**Q8**
**Drop all the samples(rows) with missing values strictly greater than 2. How many samples remains after that ? (Note: compute after removing "?")**


In [17]:
df_cleaned = df.dropna(thresh=df.shape[1] - 2)
df_cleaned.shape[0]


9917

##**Q9**
**Drop all the samples(rows) with missing values in the original dataframe. How many samples remains after that ? (Note: compute after removing "?")**


In [18]:
df_cleaned_all_rows = df.dropna()
df_cleaned_all_rows.shape[0]

5449

#**W1-GA1.2**

##**Q1**
**Select all the even indexed rows from the given dataset. What is the value in the 3rd indexed column of the 0th indexed row in the selected dataframe. (row/ column indexed from zero(0) in python/pandas)**

In [19]:
even_index_row=df.iloc[::2]
even_index_row.head()

Date  Year    Locality  Estimated Value  Sale Price       Property  \
0  2009-01-02  2009   Waterbury         111440.0    185000.0  Single Family   
2  2009-01-02  2009   Waterbury          55720.0    140000.0  Single Family   
4  2009-01-02  2009  Bridgeport         112351.0    210000.0            NaN   
6  2009-01-02  2009     Norwalk         571060.0   1000000.0  Single Family   
8  2009-01-02  2009         NaN         603540.0   1680000.0  Single Family   

      Residential  num_rooms  num_bathrooms  carpet_area  property_tax_rate  \
0  Detached House          3              3        996.0           1.025953   
2  Detached House          3              2        910.0           1.025953   
4  Detached House          3              2       1092.0           1.025953   
6  Detached House          3              2          NaN           1.025953   
8  Detached House          3              1       1039.0           1.025953   

    Face  
0  South  
2  South  
4   East  
6  South  
8   East

In [20]:
even_index_row.iloc[0,3]

111440.0

##**Q2**
**Select all the even indexed rows from the given dataset. What is the value in the 3rd indexed column of the 332th indexed row in the selected dataframe. (row/ column indexed from zero(0))**


In [21]:
even_index_row.iloc[332,3]

256200.0

##**Q3**
**Select all the odd indexed rows from the given dataset. What is the value in the 3rd indexed column of the 332th indexed row in the selected dataframe. (row/ column indexed from zero(0))
 Please consider case sensitivity while giving your solution capital letter ≠ small letter
  If it is a string don't write quotes in the solution.**



In [22]:
odd_index_row=df.iloc[1::2]
odd_index_row.head()

Date  Year    Locality  Estimated Value  Sale Price       Property  \
1  2009-01-02  2009  Bridgeport         124670.0    150000.0     Two Family   
3  2009-01-02  2009  Bridgeport        4775276.0    272900.0  Single Family   
5  2009-01-02  2009   Greenwich        2868880.0   3300000.0  Single Family   
7  2009-01-02  2009   Greenwich         592760.0    920000.0  Single Family   
9  2009-01-02  2009  Bridgeport         137901.0    245000.0  Single Family   

      Residential  num_rooms  num_bathrooms  carpet_area  property_tax_rate  \
1          Duplex          4              3       1241.0           1.025953   
3  Detached House          3              1        971.0           1.025953   
5  Detached House          3              1       1039.0           1.025953   
7  Detached House          3              3        937.0           1.025953   
9  Detached House          3              1        956.0           1.025953   

    Face  
1  South  
3   East  
5  North  
7  North  
9   West

In [23]:
odd_index_row.iloc[332,3]

352940.0

##**Q4**
**Select all the odd indexed columns from the given dataset. What is the value in the 5th indexed column of the 100th indexed row in the selected dataframe. (row/ column indexed from zero(0))
Please consider case sensitivity while giving your solution capital letter ≠ small letter
If it is a string don't write quotes in the solution.**

In [24]:
odd_column=df.iloc[:,1::2]
odd_column.iloc[100,5]

'West'

##**Q5**
**Select all the even indexed columns and even indexed rows from the given dataset. What is the value in the 3th indexed column of the 255th indexed row in the selected dataframe. (row/ column indexed/starts from zero(0))
Please consider case sensitivity while giving your solution capital letter ≠ small letter
If it is a string don't write quotes in the solution.**

In [25]:
even_rows_columns = df.iloc[::2, ::2]

value = even_rows_columns.iloc[255, 3]
value


'Detached House'

##**Q6**

**Determine the number of samples (rows) in the dataset that belong to the six most recent years, based on the provided data.**

In [26]:
max_year = df['Year'].max()
recent_years_count = df.loc[df['Year'].isin(range(max_year - 5, max_year + 1))].shape[0]
print(recent_years_count)


4749


##**Q7**
**Determine the number of samples (rows) in the dataset which has num_rooms = 3 and num_bathrooms = 3**

In [28]:
sample1= df[(df['num_rooms'] == 3) & (df['num_bathrooms'] == 3)]
sample1.shape[0]

2730

    In Python (and Pandas), the & operator is used for element-wise logical AND operations between two conditions, often when filtering a DataFrame.
    It is commonly used inside a Pandas operation to combine multiple conditions.
    Always wrap each condition in parentheses to avoid precedence issues and ensure the expression works as expected.

##**Q8**
**Determine the number of samples (rows) in the dataset which has num_rooms = 3 or num_bathrooms = 3**

In [29]:
sample2= df[(df['num_rooms'] == 3) | (df['num_bathrooms'] == 3)]
sample2.shape[0]

8591

    The | operator in Pandas is used for element-wise logical OR operations between two conditions.
    It allows you to filter rows in a DataFrame where at least one of the conditions is True.
    Always use parentheses around each condition to avoid precedence issues.







##**Q9**
**In the Year 2022, how many houses (rows) located in the Greenwich Locality have exactly num_room == 3, and facing towards either the North or East?**

In [30]:
sample3 = df[(df['Year'] == 2022) &
              (df['Locality'] == 'Greenwich') &
              (df['num_rooms'] == 3) &
              ((df['Face'] == 'North') | (df['Face'] == 'East'))]
sample3.shape[0]

42

##**Q10**
**Calculate the number of samples(rows) that correspond to the month of August across all the years.?**

In [31]:
sample4 = df[df['Date'].dt.month==8]
sample4.shape[0]

1050

    dt.month: It is used to extract the month from a datetime column or series.
    This attribute is available when a column or Series has been converted to the datetime format.
        Purpose: Extracts the month (as an integer) from a datetime object.
        Usage: It returns an integer between 1 and 12, representing the month.

##**Q11**
**According to the data which Locality have highest average Sale Price?**

In [32]:
loc= df.groupby('Locality')['Sale Price'].mean()
print(loc.idxmax(),loc.max())

Greenwich 2270223.539495238
